apps/aestratum/src/aestratum_job.erl

```diff
@@ -0,0 +1,204 @@
+    MaxSolveTime.
+
+-spec created(job()) -> created().
+created(#job{created = Created}) ->
+    Created.
+
+-spec shares(job()) -> [share()].
+shares(#job{shares = Shares}) ->
+    Shares.
+
+-spec error_shares(job()) -> [error_share()].
+error_shares(#job{error_shares = ErrorShares}) ->
+    ErrorShares.
+
+-spec is_submitted(job()) -> boolean().
+is_submitted(#job{shares = Shares}) when Shares =/= [] ->
+    true;
+is_submitted(#job{shares = []}) ->
+    false.
+
+-spec solve_time(job()) -> solve_time().
+solve_time(#job{created = Created, shares = Shares}) when Shares =/= [] ->
+    %% The shortest time is considered to be the solve time. The first
+    %% submitted share, which has the shortest solve time, is the last in
+    %% the list.
+    Share = lists:last(Shares),
+    aestratum_share:created(Share) - Created;
+solve_time(#job{max_solve_time = MaxSolveTime, shares = []}) ->
+    MaxSolveTime.
+
+-spec make_id(block_hash(), block_version(), block_target()) -> term(). %% TODO
+make_id(BlockHash, BlockVersion, BlockTarget) ->
+    make_id1(BlockHash, BlockVersion, BlockTarget).
+
+-spec status(job()) -> map().
+status(#job{id = Id,
+            block_hash = BlockHash,
+            block_version = BlockVersion,
+            block_target = BlockTarget,
+            share_target = ShareTarget,
+            desired_solve_time = DesiredSolveTime,
+            max_solve_time = MaxSolveTime,
+            shares = Shares,
+            error_shares = ErrorShares,
+            created = Created}) ->
+    #{id => Id,
+      block_hash => BlockHash,
+      block_version => BlockVersion,
+      block_target => aestratum_target:to_hex(BlockTarget),
+      share_target => aestratum_target:to_hex(ShareTarget),
+      desired_solve_time => DesiredSolveTime,
+      max_solve_time => MaxSolveTime,
+      shares => [aestratum_share:status(S) || S <- Shares],
+      error_shares => [aestratum_share:status(S) || S <- ErrorShares],
+      created => Created}.
+
+-spec add_share(share(), job()) -> job().
+add_share(Share, #job{shares = Shares} = Job) ->
+    %% TODO: add limit on how many shares can be submitted per job.
+    Job#job{shares = [Share | Shares]}.
+
+-spec add_error_share(error_share(), job()) -> job().
+add_error_share(ErrorShare, #job{error_shares = ErrorShares} = Job) ->
+    %% TODO: add limit on how many error shares can be submitted per job.
+    Job#job{error_shares = [ErrorShare | ErrorShares]}.
+
+-spec is_share_present(miner_nonce(), pow(), job()) -> boolean().
+is_share_present(MinerNonce, Pow, #job{shares = Shares}) ->
+    lists:any(
+      fun(Share) ->
+              aestratum_share:is_duplicate(MinerNonce, Pow, Share)
+      end, Shares).
+
+%% Internal functions.
+
+make_id1(BlockHash, BlockVersion, BlockTarget) ->
+    BlockTarget1 = aestratum_target:to_hex(BlockTarget),
+    <<Id:8/binary, _Rest/binary>> =
+        crypto:hash(sha256, [BlockHash, BlockTarget1, BlockVersion]),
+    to_hex(Id).
+
+to_hex(Bin) ->
```

Comment:
Use `aeu_hex`?

Reply:
Not the same result - binary vs list and downcased vs upcased....